### PR TITLE
feat: add a context menu option to translate the selected text

### PR DESCRIPTION
**Pull request by shinglyu**
This adds a context menu option for you to translate the selected text. 
Screenshot 2022-03-03 at 15 01 47
Screenshot 2022-03-03 at 15 01 55

* Issue #26*  

*Description of changes:*

This is a draft PR for #26. I need to discuss with you about the UI design. Currently I'm using an `alert` for simplicity. 
Also I'm directly doing the translation in the background script. Since there is no need to pass the string to the content script and let content script translate it. Please let me know if this is acceptable. 

Also apologies in advance if I didn't comply with your coding convention, test coverage requirements, commit message conventions etc. Please kindly let me know your conventions and quality requirements and I'll try to fix them ASAP. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AJStacy** [comment]
I love the idea of have a context menu item for applying a translation. In a recent release I have added hotkey support for translating a page which also creates an indicator for the user to know a translation is occurring.

As far as popping up the translation rather than replacing the text nodes... I think I'm leaning towards the text nodes being swapped with the translated text for a few of reasons:

1. Alerts block other user interactions with the page until they are dismissed.
2. Once the alert is dismissed you would have to reselect the text and execute another translation.
3. Is this translation text cached? The process in the contentScript caches the results to reduce the API load on re-translations.
4. There may be cases where the user wants to print the webpage and if the text is in another context it could cause strange behaviors with rendering the document for printing.

I'll leave some comments on your commit to point out any code quality or pattern issues that I find.

Thank you for being willing to contribute!

**Comment by shinglyu** [comment]
How about I create two options in the context menu: "Translate in-place" and "Translate in a popup" (we can adjust the wording)? For complex pages, I'm afraid the problem with #17 would happen and the user gets stuck. Also for RTL languages like Arabic or Hebrew, I'm afraid if we translate them to a LTR language, it won't fit into the existing page layout. A popup would be a good fallback for pages that we can't replace in-place properly.

Google translate extension solve this in two ways:
1. When you select a text, a small icon appears next to the text. Click on it will open a small text bubble next to the text. 
2. If you click on the extension icon in the browser menu bar. The content of the popup will become the translation.

Yes the alert is just an POC. I can build a nicer looking popup using HTML/CSS and preserve the text selection and does not block the browser UI interaction.
I'll also work on the caching and address your comments once I get a go-ahead for the proposed design :)

**Comment by AJStacy** [comment]
Responses to each point:

1. Having two context menu options and two methods of returning the interpretation is a good compromise. We might want to update this PR to focus on the "Translate in Popup" feature and then open a new ticket for the "Translate in Place" feature.
2. Your point about RTL to LTR is one I didn't consider and definitely makes the "Translate in Popup" feature more valuable.
3. I assumed the alert was just a POC but wanted to be clear. :)

I'll be awaiting the proposed UI proof of concept. This is leading me down another trail of thought... what if we could allow the user to translate pieces of a page and collect them in a document in memory. Once the user is done, they could export the document to their computer or via some other transport mechanism. Thoughts?

**Comment by shinglyu** [comment]
Hi @AJStacy, sorry for the late reply. I was out sick for about 3 weeks, then work piled up so I was busy digesting them. 

Yes I'd love to focus this PR on the Transate in Popup and open a follow up issue on Translate in Place. I'll also open a followup issue on the "store translations in memory and export at once" feature. 

I'll try to update this PR to have a in-page popup using JavaScript. It might take a few more days as I'm slowly getting back on my feet after the sick leave.

**Comment by AJStacy** [comment]
That looks good. As far as the optimizations:
1. There is already a "translating..." indicator that is used for the keyboard shortcut for translations that you should be able to use.
2. Yes
3. What if we used a web component here for rendering the text? The shadow DOM would ensure that no CSS leaks into the overlay styles.

**Comment by shinglyu** [comment]
@AJStacy I've implemented 1 and 2. Would you mind if I leave the Web Component part as a followup PR? I've low on bandwidth and I don't want to block you on this PR forever.

**Comment by shinglyu** [comment]
Looks like using WebComponents in content script requires a [polyfill](https://stackoverflow.com/questions/42800035/why-cant-you-create-custom-elements-in-content-scripts). Do you think it's worth adding the dependency? The license for the polyfill is BSD and I'm not sure if that will be an issue.

**Comment by AJStacy** [comment]
Oh! I was unaware that a polyfill was required for it. I don't want to add any more dependencies or polyfills to the project so I guess we'll have to go with a reset CSS stylesheet to make it work.

**Review by AJStacy — APPROVED** [review]

File: src/background/main.ts [hunk 1]
```diff
@@ -117,3 +119,38 @@ onMessage('get-current-tab', async () => {
     };
   }
 });
+
+// Translate selection with right-click menu
+const escape = (text: string): string => {
+  return text.replaceAll('"', '\\"').replaceAll("'", "\\'");
+};
+
+function translateSelectionHandler() {
```
[the remark this answers is not in the document]
Comment (shinglyu) [reply src/background/main.ts:128]: Sorry, which comment do you mean exactly? About splitting "translate in popup" and "translate inplace" or about caching?

**Comment by shinglyu** [comment]
Hi @AJStacy I'm not sure I understand you last comment. How should we move this forward?

**Comment by AJStacy** [comment]
I went ahead and merged it. I'm not sure what my comment was about 😅

**Pull request merged**